cmd/tailscale/cli/qr.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package cli
+
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"syscall"
+	"unsafe"
+
+	"github.com/mattn/go-isatty"
+	qrcode "github.com/skip2/go-qrcode"
+	"golang.org/x/sys/unix"
+)
+
+// --qr-format values
+const (
+	qrFormatAuto  = "auto"
+	qrFormatASCII = "ascii"
+	qrFormatLarge = "large"
+	qrFormatSmall = "small"
+)
+
+// FormatQRCode returns a QR code for s.
+func formatQRCode(w io.Writer, format, s string) (string, error) {
+	const inverse = false
+
+	q, err := qrcode.New(s, qrcode.Medium)
+	if err != nil {
+		return "", fmt.Errorf("QR code error: %w", err)
+	}
+
+	if format == qrFormatAuto {
+		// Assume that the full UTF-8 character set is available.
+		format = qrFormatSmall
+
+		switch effectiveGOOS() {
+		case "linux":
+			format, err = qrFormatLinux(w, inverse)
+			if err != nil {
+				return "", fmt.Errorf("QR code error: %w", err)
+			}
+		}
+	}
+
+	switch format {
+	case qrFormatASCII:
+		s := q.ToString(inverse)
+		return strings.ReplaceAll(s, "█", "#"), nil
+	case qrFormatLarge:
+		return q.ToString(inverse), nil
+	case qrFormatSmall:
+		return q.ToSmallString(inverse), nil
+	default:
+		return "", fmt.Errorf("unknown QR code format: %q", format)
+	}
+}
+
+func qrFormatLinux(w io.Writer, inverse bool) (format string, _ error) {
+	// Almost every terminal supports UTF-8, but the Linux
+	// console may have partial or no support, which is
+	// especially painful inside VMs. See #12935.
+	format = qrFormatSmall
+
+	// Is the locale (LC_CTYPE) set to UTF-8?
+	locale, err := LinuxLocale()
+	if err != nil {
+		return qrFormatASCII, fmt.Errorf("QR: %w", err)
+	}
+	const utf8 = ".UTF-8"
+	if !strings.HasSuffix(locale["LC_CTYPE"], utf8) &&
+		!strings.HasSuffix(locale["LANG"], utf8) {
+		return qrFormatASCII, nil
+	}
+
+	// Are we printing to a terminal?
+	f, ok := w.(*os.File)
+	if !ok {
+		return format, nil
+	}
+	if !isatty.IsTerminal(f.Fd()) {
+		return format, nil
+	}
+	fd := f.Fd()
+
+	// On a Linux console, check that the current keyboard
+	// is in Unicode mode. See unicode_start(1).
+	const K_UNICODE = 0x03
+	kbMode, err := LinuxIoctlGetKBMode(fd)
+	if err != nil {
+		if errors.Is(err, syscall.ENOTTY) {
+			return format, nil
+		}
+		return "", err
+	}
+	if kbMode != K_UNICODE {
+		return qrFormatASCII, nil
+	}
+
+	// On a raw Linux console, detect whether the block
+	// characters are available in the current font by
+	// consulting the Unicode-to-font mapping.
+	unimap, err := LinuxIoctlGetUniMap(fd)
+	if err != nil {
+		if errors.Is(err, syscall.ENOTTY) {
+			return format, err
+		}
+		return "", err
+	}
+	if _, ok := unimap['█']; ok {
+		format = qrFormatLarge
+	}
+	if _, ok := unimap['▀']; ok && inverse {
+		format = qrFormatSmall
+	}
+	if _, ok := unimap['▄']; ok && !inverse {
+		format = qrFormatSmall
+	}
+
+	return format, nil
+}
+
+func LinuxLocale() (map[string]string, error) {
```

Comment:
did you mean to export this? weird place if so.

docs on that map's keys/values in any case

Reply:
Nope, my fingers capitalize Linux automatically. Unexported.